tfx_addons/sampler/executor.py

```diff
@@ -0,0 +1,259 @@
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Executor for TensorFlow Transform."""
+
+import os
+import random
+from typing import Any, Dict, List, Text
+
+import apache_beam as beam
+import tensorflow as tf
+from tfx import types
+from tfx.dsl.components.base import base_beam_executor
+from tfx.dsl.io import fileio
+from tfx.types import artifact_utils
+from tfx.utils import io_utils, json_utils
+
+from tfx_addons.sampler import spec
+
+
+class Executor(base_beam_executor.BaseBeamExecutor):
+  """Executor for Sampler."""
+  def _CreatePipeline(self, unused_transform_output: Text) -> beam.Pipeline:
+    """Creates beam pipeline.
+    Args:
+      unused_transform_output: unused.
+    Returns:
+      Beam pipeline.
+    """
+
+    return self._make_beam_pipeline()
+
+  def Do(
+      self,
+      input_dict: Dict[Text, List[types.Artifact]],
+      output_dict: Dict[Text, List[types.Artifact]],
+      exec_properties: Dict[Text, Any],
+  ) -> None:
+    """Sampler executor entrypoint.
+    Args:
+      input_dict: Input dict from input key to a list of artifacts, including:
+      - examples: A list of type `standard_artifacts.Examples` which should
+      contain custom splits specified in splits_config. If custom split is
+      not provided, this should contain two splits 'train' and 'eval'.
+      output_dict: Output dict from key to a list of artifacts, including:
+      - sampled_examples: sampled examples, only for the given
+      splits as specified in splits. May also include copies of the
+      other non-sampled spits, as specified by null_classes.
+      exec_properties: A dict of execution properties, including:
+      - name: Optional unique name. Necessary if multiple components are
+      declared in the same pipeline.
+      - label: The name of the column containing class names to sample by.
+      - splits: A list containing splits to sample. Defaults to ['train'].
+      - copy_others: Determines whether we copy over the splits that aren't
+      sampled, or just exclude them from the output artifact. Defualts
+      to True.
+      - shards: The number of files that each sampled split should
+      contain. Default 0 is Beam's tfrecordio function's default.
+      - null_classes: A list determining which classes that we should
+      not sample. Defaults to None.
+    Returns:
+      None
+    """
+
+    self._log_startup(input_dict, output_dict, exec_properties)
+
+    label = exec_properties[spec.SAMPLER_LABEL_KEY]
```

Comment:
I'm not sure the TFX philosophy on this -- so I would read other executor classes to see the pattern there -- but is it worth protecting against bad data passed to `Do` (particularly in `exec_properties`)? Do we want to do any sort of validation on this fairly public method or just trust the caller and let things break downstream if the caller passes garbage? I think whatever other common executors do is fine, just a question I'm having

Reply:
I'm seeing more validations being put in other executors as TFX becoming more stable. We should try to catch detectable errors early because failing after all the overhead of scheduling and pre processing slows down the iteration speed significantly.

Reply:
What kinds of errors would I be catching, and what kind of data validation would I be using - TFDV?

Reply:
I was thinking simpler then that: just validating the input variables to this function (not validating the actual data itself).

There are really simple validation concerns like: what if `copy_others` is not in the dict or what if `splits` isn't json? We can probably ignore that kind of validation because the error raised will be immediate and readable when we try to index into the dict or deserialize the json. But I am interested in understanding if there are other kinds of bad configuration that could get passed to this function that would break downstream in a scheduled job in a more painful way. For example: what if `null_classes` is garbage input like a list of ints... do we just pass that all the way through? Where does it break?